ledger/artifactmanager/artifactmanager_senders.go

```diff
@@ -30,127 +30,123 @@ import (
-			return nil, err
-		}
-		castedResp, ok := response.(*reply.Code)
-		if !ok {
+			entry.reply = castedResp
 			return response, err
 		}
-
-		entry.code = castedResp
-		return response, err
 	}
 }
 
-func (m *ledgerArtifactSenders) followRedirectSender(sender Sender) Sender {
-	return func(ctx context.Context, msg core.Message, options *core.MessageSendOptions) (core.Reply, error) {
-		inslog := inslogger.FromContext(ctx)
-		inslog.Debug("LedgerArtifactManager.SendAndFollowRedirectSender starts ...")
-
-		rep, err := sender(ctx, msg, options)
-		if err != nil {
-			return nil, err
-		}
-
-		if r, ok := rep.(core.RedirectReply); ok {
-			stats.Record(ctx, statRedirects.M(1))
-
-			redirected := r.Redirected(msg)
-			inslog.Debugf("redirect reciever=%v", r.GetReceiver())
+func followRedirectSender(bus core.MessageBus) PreSender {
+	return func(sender Sender) Sender {
+		return func(ctx context.Context, msg core.Message, options *core.MessageSendOptions) (core.Reply, error) {
+			inslog := inslogger.FromContext(ctx)
+			inslog.Debug("LedgerArtifactManager.SendAndFollowRedirectSender starts ...")
 
-			rep, err = m.bus(ctx).Send(ctx, redirected, &core.MessageSendOptions{
-				Token:    r.GetToken(),
-				Receiver: r.GetReceiver(),
-			})
+			rep, err := sender(ctx, msg, options)
 			if err != nil {
 				return nil, err
 			}
-			if _, ok := rep.(core.RedirectReply); ok {
-				return nil, errors.New("double redirects are forbidden")
+
+			if r, ok := rep.(core.RedirectReply); ok {
+				stats.Record(ctx, statRedirects.M(1))
+
+				redirected := r.Redirected(msg)
+				inslog.Debugf("redirect reciever=%v", r.GetReceiver())
+
+				rep, err = bus.Send(ctx, redirected, &core.MessageSendOptions{
+					Token:    r.GetToken(),
+					Receiver: r.GetReceiver(),
+				})
+				if err != nil {
+					return nil, err
+				}
+				if _, ok := rep.(core.RedirectReply); ok {
+					return nil, errors.New("double redirects are forbidden")
+				}
+				return rep, nil
 			}
-			return rep, nil
-		}
 
-		return rep, err
+			return rep, err
+		}
 	}
 }
 
-func (m *ledgerArtifactSenders) retryJetSender(sender Sender) Sender {
-	return func(ctx context.Context, msg core.Message, options *core.MessageSendOptions) (core.Reply, error) {
-		inslog := inslogger.FromContext(ctx)
-		inslog.Debug("LedgerArtifactManager.RetryJetSender starts ...")
-
-		retries := jetMissRetryCount
+func retryJetSender(pulseStorage core.PulseStorage, jetStorage storage.JetStorage) PreSender {
+	return func(sender Sender) Sender {
+		return func(ctx context.Context, msg core.Message, options *core.MessageSendOptions) (core.Reply, error) {
+			inslog := inslogger.FromContext(ctx)
+			inslog.Debug("LedgerArtifactManager.RetryJetSender starts ...")
 
-		currentPulse, err := m.pulseStorage.Current(ctx)
-		if err != nil {
-			return nil, err
-		}
+			retries := jetMissRetryCount
 
-		for retries > 0 {
-			rep, err := sender(ctx, msg, options)
+			currentPulse, err := pulseStorage.Current(ctx)
```

Comment:
We need to pass PulseNumber here, not PulseStorage. When conveyer comes, all calls to PulseStorage would be forbidden.